### PR TITLE
Upload n types

### PR DESCRIPTION
**Pull request by pldespaigne**
Remove the crappy stuff from #4836

### CODE REVIEW

File: apps/backend-functions/src/media.ts [hunk 1]
```diff
@@ -106,11 +109,23 @@ export async function linkFile(data: storage.ObjectMetadata) {
...
-    // because of possible nested map and arrays, we need to retrieve the whole document
-    // modify it, then update the whole doc with the new (modified) version
 
-    // TODO issue#4002 refactor as set(doc, metadata.field, { ref: finalPath});
-    set(doc, metadata.field, finalPath); // update the whole doc with only the new ref
+    // separate extraData from metadata
+    const extraData = {...metadata};
+    const keysToDelete = [
+      'uid', 'privacy', 'collection', 'docId', 'field',
+      'firebaseStorageDownloadTokens',
```
Comment (pldespaigne) [inline apps/backend-functions/src/media.ts:117]: remove only `firebaseStorageDownloadTokens` & `uid`

File: apps/backend-functions/src/media.ts [hunk 2]
```diff
@@ -106,11 +109,23 @@ export async function linkFile(data: storage.ObjectMetadata) {
     await file.copy(to);
 
     // (3) update db
-    // because of possible nested map and arrays, we need to retrieve the whole document
-    // modify it, then update the whole doc with the new (modified) version
 
-    // TODO issue#4002 refactor as set(doc, metadata.field, { ref: finalPath});
-    set(doc, metadata.field, finalPath); // update the whole doc with only the new ref
+    // separate extraData from metadata
+    const extraData = {...metadata};
```
Comment (pldespaigne) [inline apps/backend-functions/src/media.ts:114]: do not copy

File: apps/backend-functions/src/media.ts [hunk 3]
```diff
@@ -342,30 +354,34 @@ export async function cleanOrgMedias(before: OrganizationDocument, after?: Organ
 }
 
 export async function cleanMovieMedias(before: MovieDocument, after?: MovieDocument): Promise<void> {
+
+  const needsToBeCleaned = (beforeStoragePath: string | undefined, afterStoragePath: string) => {
+    return !!beforeStoragePath && (beforeStoragePath !== afterStoragePath && afterStoragePath === '')
```
Comment (pldespaigne) [inline apps/backend-functions/src/media.ts:359]: no prenthesis

File: libs/media/src/lib/+state/media.firestore.ts [hunk 4]
```diff
@@ -1,16 +1,18 @@
 
-import { privacies, Privacy } from "@blockframes/utils/file-sanitizer";
+import { FileMetaData } from './media.model';
+
+/**
+* Representation of a storage file in our Firestore db.
+* @note this is not the same as the data needed to upload into the storage
+*/
+export interface StorageFile {
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.firestore.ts:8]: New interface with every fields

File: libs/media/src/lib/+state/media.model.ts [hunk 5]
```diff
@@ -91,26 +129,30 @@ export function recursivelyListFiles(document: any): string[] {
   } else if (Array.isArray(document)) {
     const result = document.map(el => recursivelyListFiles(el));
     // pre-ES2019 Array flattening, with ES2019 we could use Array.prototype.flat()
-    return [].concat(...result);
+    // return [].concat(...result);
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.model.ts:132]: clean

File: libs/media/src/lib/+state/media.model.ts [hunk 6]
```diff
@@ -91,26 +129,30 @@ export function recursivelyListFiles(document: any): string[] {
...
+    return result.flat();
 
   } else if (!document) {
     return [];
 
   } else if (typeof document === 'object') {
     const result = Object.keys(document).map(key => recursivelyListFiles(document[key]));
     // pre-ES2019 Array flattening, with ES2019 we could use Array.prototype.flat()
-    return [].concat(...result);
+    // return [].concat(...result);
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.model.ts:141]: clean

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  public fileName: string;
+
+  constructor(
+    private snackBar: MatSnackBar,
+    private uploaderService: FileUploaderService,
+  ) { }
+
+  ngOnInit() {
+
+    console.log(this.displayFile); // TODO REMOVE DEBUG LOG
```
Comment (pldespaigne) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts:88]: clean

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+    this.computeState();
+  }
+
+  ngOnChanges(changes: SimpleChanges) { // TODO REMOVE DEBUG LOG
```
Comment (pldespaigne) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts:93]: clean

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  @HostListener('dragleave', ['$event'])
+  onDragLeave($event: DragEvent) {
+    $event.preventDefault();
+    this.computeState();
+  }
+
+
+  public selected(files: FileList | File) {
+
+    console.log('selected', files, this.metadata); // TODO REMOVE DEBUG LOG
```
Comment (pldespaigne) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts:119]: clean

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+   * @note it **should not** start with `tmp/`
+   * @note it **should not** start with a `/`
+   * @note it **should not** end with a `/`
+   * @example 'protected/movie/1234/poster'
+   */
+  @Input() storagePath: string;
+
+  @Input() metadata: FileMetaData;
+
+  @Input() displayFile: FormControl;
```
Comment (pldespaigne) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.ts:58]: rename into `form`

File: libs/movie/src/lib/movie/+state/movie.model.ts [hunk 11]
```diff
@@ -112,12 +112,12 @@ export function createMoviePromotional(
     ...params,
     clip_link: params.clip_link ?? '',
     financialDetails: params.financialDetails ?? '',
-    moodboard: params.moodboard ?? '',
+    moodboard: params.moodboard ?? { storagePath: '' },
```
Comment (pldespaigne) [inline libs/movie/src/lib/movie/+state/movie.model.ts:115]: use a create function + new interface

File: libs/movie/src/lib/movie/form/media-files/media-files.component.ts [hunk 12]
```diff
@@ -12,6 +17,38 @@ import { DynamicTitleService } from '@blockframes/utils/dynamic-title/dynamic-ti
...
+      uid: '',
+    },
+    moodboard: {
+      privacy: 'public',
+      collection: 'movies',
+      docId: this.movieId,
+      field: 'promotional.moodboard',
+      uid: '',
+    },
+  };
```
Comment (pldespaigne) [inline libs/movie/src/lib/movie/form/media-files/media-files.component.ts:50]: ok but find a better way, maybe extract it

File: libs/movie/src/lib/movie/form/media-files/media-files.component.ts [hunk 13]
```diff
@@ -12,6 +17,38 @@ import { DynamicTitleService } from '@blockframes/utils/dynamic-title/dynamic-ti
 export class MovieFormMediaFilesComponent {
   form = this.shell.getForm('movie');
 
+  movieId = this.route.snapshot.params.movieId;
+
+  storagePaths: Record<string, string> = {
+    presentationDeck: `public/movies/${this.movieId}/presentationDeck`,
+    scenario: `public/movies/${this.movieId}/scenario`,
+    moodboard: `public/movies/${this.movieId}/moodboard`,
```
Comment (pldespaigne) [inline libs/movie/src/lib/movie/form/media-files/media-files.component.ts:25]: don't modify the path right now

File: libs/movie/src/lib/movie/form/movie.form.ts [hunk 14]
```diff
@@ -654,10 +654,23 @@ function createMoviePromotionalElementsControls(movieId: string, promotionalElem
 
     // Hosted Media
     financialDetails: new HostedMediaForm(entity.financialDetails, { privacy: 'public', collection: 'movies', docId: movieId ?? '', field: 'promotional.financialDetails'}),
-    presentation_deck: new HostedMediaForm(entity.presentation_deck, { privacy: 'public', collection: 'movies', docId: movieId ?? '', field: 'promotional.presentation_deck'}),
-    scenario: new HostedMediaForm(entity.scenario, { privacy: 'public', collection: 'movies', docId: movieId ?? '', field: 'promotional.scenario'}),
-    moodboard: new HostedMediaForm(entity.moodboard, { privacy: 'public', collection: 'movies', docId: movieId ?? '', field: 'promotional.moodboard'}),
-    notes: FormList.factory(entity.notes, (el, i) => new MovieNotesForm(el, { privacy: 'public', collection: 'movies', docId: movieId ?? '', field: `promotional.notes[${i}].ref` })),
+    presentation_deck: new FormEntity({
```
Comment (pldespaigne) [inline libs/movie/src/lib/movie/form/movie.form.ts:657]: use a create function

File: libs/media/src/lib/+state/static-files.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  // campaigns
+  'budget',
+  'financingPlan',
+  'waterfall',
+] as const;
+export type FileLabel = typeof fileLabels[number];
+
+export function getFileStoragePath(collection: CollectionHoldingFile, label: FileLabel) {
+  const storagePaths: Record<CollectionHoldingFile, Partial<Record<FileLabel, string>>> = {
+    'orgs': {
```
Comment (GrandSchtroumpf) [inline libs/media/src/lib/+state/static-files.ts:40]: remove quotes for keys that are not kebab-case

File: libs/media/src/lib/+state/static-files.ts [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+  // campaigns
+  'budget',
+  'financingPlan',
+  'waterfall',
+] as const;
+export type FileLabel = typeof fileLabels[number];
+
+export function getFileStoragePath(collection: CollectionHoldingFile, label: FileLabel) {
+  const storagePaths: Record<CollectionHoldingFile, Partial<Record<FileLabel, string>>> = {
```
Comment (GrandSchtroumpf) [inline libs/media/src/lib/+state/static-files.ts:39]: Shouldn't this be outside of the function ?

File: libs/media/src/lib/+state/static-files.ts [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+  const pathPart = storagePaths[collection][label];
+  if (!pathPart) return '';
+
+  return `${collection}/${pathPart}`;
+}
+
+export function getFileMetadata(collection: CollectionHoldingFile, label: FileLabel, docId: string, index?: number) {
+  const metadatas: Record<CollectionHoldingFile, Partial<Record<FileLabel, FileMetaData>>> = {
+    'orgs': {
```
Comment (GrandSchtroumpf) [inline libs/media/src/lib/+state/static-files.ts:71]: Same as above

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.scss [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+section {
+  height: 100%;
+  width: 100%;
+  padding: 24px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  flex-direction: column;
+  border: solid 1px var(--foreground-divider);
+  border-radius: 5px;
```
Comment (GrandSchtroumpf) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.scss:20]: 4px

File: libs/media/src/lib/file/file-new-uploader/file-uploader.component.scss [hunk 19]
```diff
@@ -0,0 +1,75 @@
+:host {
+  height: 100%;
+  width: 100%;
+}
+
+img.picture-hover {
+  height: 142px;
```
Comment (GrandSchtroumpf) [inline libs/media/src/lib/file/file-new-uploader/file-uploader.component.scss:7]: 140px

File: libs/movie/src/lib/movie/form/media-files/media-files.component.html [hunk 20]
```diff
@@ -3,28 +3,47 @@
 
     <section id="presentation-deck">
       <h3>Presentation Deck</h3>
-      <file-uploader test-id="presentation-deck" [form]="form.promotional.get('presentation_deck')"
-        [storagePath]="getPath('presentation_deck')" allowedFileType="pdf">
+
+      <file-new-uploader
+        test-id="presentationDeck"
+        accept="pdf"
+        [form]="form.promotional.get('presentation_deck').get('storagePath')"
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/media-files/media-files.component.html:10]: I think in the future, this should be the `presentation_deck` itself.

File: libs/movie/src/lib/movie/form/movie.form.ts (hunk 14 above)
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/movie.form.ts:657]: Maybe create a `StorageFileForm`

**Review by GrandSchtroumpf — APPROVED** [review]

**Pull request merged**